flatland/envs/agent_chains.py

```diff
@@ -119,99 +117,78 @@ def find_stop_preds(self, svStops=None):
         """ Take a list of stopped agents, and apply a stop color to any chains/trees
             of agents trying to head toward those cells.
-            Count the number of agents blocked, ignoring those which are already marked.
-            (Otherwise it can double count swaps)
+            :param svStops: list of stopped agents
+            :param color: color to apply to predecessor of stopped agents
+            :return: all predecessors of any stopped agent
 
         """
-        iCount = 0
-        svBlocked = set()
+        predecessors = set()
         if len(svStops) == 0:
-            return svBlocked
+            return predecessors
 
         # The reversed graph allows us to follow directed edges to find affected agents.
         Grev = self.get_G_reversed()
         for v in svStops:
 
             # Use depth-first-search to find a tree of agents heading toward the blocked cell.
             lvPred = list(nx.traversal.dfs_postorder_nodes(Grev, source=v))
-            svBlocked |= set(lvPred)
-            svBlocked.add(v)
-            # print("node:", v, "set", svBlocked)
-            # only count those not already marked
+            predecessors |= set(lvPred)
+            predecessors.add(v)
+
+            # only color those not already marked (not updating previous colors)
             for v2 in [v] + lvPred:
                 if self.G.nodes[v2].get("color") != color:
                     self.G.nodes[v2]["color"] = color
-                    iCount += 1
-
-        return svBlocked
+        return predecessors
 
     def find_conflicts(self):
-        self.reset_G_reversed()
+        """Called in env.step() before the agents execute their actions."""
 
-        svStops = self.find_stops2()  # voluntarily stopped agents - have self-loops
-        # svSwaps = self.find_swaps()   # deadlocks - adjacent head-on collisions
-        svSwaps = self.find_swaps2()  # faster version of find_swaps
+        svStops: Set[Cell] = self.find_stopped_agents()  # voluntarily stopped agents - have self-loops ("same cell to same cell")
+        svSwaps: Set[Cell] = self.find_swaps()  # deadlocks - adjacent head-on collisions
 
-        # Block all swaps and their tree of predessors
-        self.svDeadlocked = self.block_preds(svSwaps, color="purple")
+        # Mark all swaps and their tree of predecessors with purple - these are directly deadlocked
+        self.svDeadlocked: Set[Cell] = self.mark_preds(svSwaps, color="purple")
 
-        # Take the union of the above, and find all the predecessors
-        # svBlocked = self.find_stop_preds(svStops.union(svSwaps))
-
-        # Just look for the tree of preds for each voluntarily stopped agent
-        svBlocked = self.find_stop_preds(svStops)
+        # Just look for the tree of preds for each voluntarily stopped agent (i.e. not wanting to move)
+        # TODO why not re-use mark_preds(swStops, color="red")?
```

Comment:
@hagrid67 @aiAdrian why do we need `find_stop_preds` - couldn't we not just re-use `mark_preds(swStops, color="red")`?

Reply:
Gosh yes, this does seem to be pointless.  

FWIW, the logic in `mark_preds` (called `block_preds` in my older version) does not match the comment "only color those not already marked" but it looks like it overwrites the color.  In my older version it seems to count the ones it has to change, which would justify the check; but then the count is discarded anyway.